zql/parser-support.go

```diff
@@ -157,14 +157,70 @@ func fieldExprArray(val interface{}) []ast.FieldExpr {
 	return ret
 }
 
-func makeSortProc(fieldsIn, dirIn, limitIn interface{}) *ast.SortProc {
-	fields := fieldExprArray(fieldsIn)
-	sortdir := dirIn.(int)
+type ProcArg struct {
+	Name  string
+	Value string
+}
+
+type ProcParam struct {
+	Present bool
+	Value   string
+}
+
+// parseArgs normalizes a list of passed-in arguments.  It takes an ordered
+// list of argument names and matches them against an array of ProcArg
+// structs which can be in any order.  It returns an array of ProcParam
+// structs, corresponding to the entries in the names array that was passed
+// in.  If any argument is provided more than once, an error is returned.
+//
+// Note that this function does not filter out invalid arguments, that is
+// assumed to be handled in the grammar.
+func parseArgs(names []string, argsIn interface{}) ([]ProcParam, error) {
+	argsArray := argsIn.([]interface{})
+	args := make([]ProcArg, len(argsArray))
+	for i, a := range argsArray {
+		args[i] = *a.(*ProcArg)
+	}
+	ret := make([]ProcParam, len(names))
+	for i, name := range names {
```

Comment:
since the arg names are hard-coded in the grammar, do we need to pass in `names` here and iterate over it?

Reply:
I'm not sure I understand the question.  The grammar validates any arguments passed to sort, but it doesn't require that they all appear or enforce an order on them.  This function is about reconciling all that.

Reply:
I guess a different way of asking my question is: do the go and js implementations need to be so different? I'm probably missing something obvious...

Reply:
Good point, I wrote the go version first and then the js version much later.  I like the js version better, I've adapted the go implementation to match.